### PR TITLE
Information Content Weighted Structural Similarity Index (IW-SSIM)

### LINKED ISSUE #128: Metric: IW-SSIM
**Is your feature request related to a problem? Please describe.**
Based on ordinary MS-SSIM this metric adds information weighting to assign different weights to different regions.

**Describe the solution you'd like**
Implementation can be based on MS-SSIM with additional weight maps

**Additional context**
Original paper: https://ece.uwaterloo.ca/~z70wang/publications/IWSSIM.pdf
MATLAB code: https://ece.uwaterloo.ca/~z70wang/research/iwssim/
Python code: https://github.com/Jack-guo-xy/Python-IW-SSIM

### LINKED ISSUE #303: MS-SSIM Tests. Bug in scale weights check.
Multi Scale SSIM uses scaling weights to compute a score for image pair. The [test 1](https://github.com/photosynthesis-team/piq/blob/19e3cb39b511cc6dbfd68ad7b564e3caa26a29b0/tests/test_ms_ssim.py#L78
) and [test 2](https://github.com/photosynthesis-team/piq/blob/19e3cb39b511cc6dbfd68ad7b564e3caa26a29b0/tests/test_ms_ssim.py#L201), which checks weight format, raises assertion error due to different purpose. 

**To Reproduce**
1. Run `pytest tests/test_ms_ssim.py` 
2. Observe no error

1. Open `test/test_ms_ssim.py`
2. Move to [line 78](https://github.com/photosynthesis-team/piq/blob/19e3cb39b511cc6dbfd68ad7b564e3caa26a29b0/tests/test_ms_ssim.py#L78)
3. See that AssetionError is raised due to missing `x` variable rather than wrong weight format.
4. Move to [line 201](https://github.com/photosynthesis-team/piq/blob/19e3cb39b511cc6dbfd68ad7b564e3caa26a29b0/tests/test_ms_ssim.py#L201)
5. See that AssetionError is raised due to missing `x` variable rather than wrong weight format.

**Suggestion**
- Fix the test
- Add assertion check, which is missing in [MS-SSIM](https://github.com/photosynthesis-team/piq/blob/master/piq/ms_ssim.py) function and class

### PR DESCRIPTION
**Pull request by denproc**
With this pull request we introduce Information Content Weighted Structural Similarity.
The measure has two interfaces as a function for metric and as class for loss function.

Closes #128, Replaces #201 

## Proposed Changes
- implementation of  [IW-SSIM](https://ece.uwaterloo.ca/~z70wang/publications/IWSSIM.pdf)
- tests
- documentation and README update
- performance estimations on TID2013, KADID10k and PIPAL
- bug fix, which closes #303

### CODE REVIEW

**Comment by denproc** [comment]
@snk4tr @zakajd , ready for review

File: piq/iw_ssim.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        scale_weights = torch.tensor([0.0448, 0.2856, 0.3001, 0.2363, 0.1333], dtype=x.dtype, device=x.device)
+    scale_weights = scale_weights / scale_weights.sum()
+    assert scale_weights.size(0) == scale_weights.numel(),\
+        f'Expected a vector of weights, got {scale_weights.dim()}D tensor'
+
+    levels = scale_weights.size(0)
+
+    min_size = (kernel_size - 1) * 2 ** (levels - 1) + 1
+    if x.size(-1) < min_size or x.size(-2) < min_size:
+        raise ValueError(f'Invalid size of the input images, expected at least {min_size}x{min_size}.')
```
Comment (zakajd) [inline piq/iw_ssim.py:84]: You test here that input satisfies requirements. 5 lines above when testing that scale weights satisfy requirements you raise an assertion, but here you raise exception. Let's be consistent.
I vote for raising assertion, because they can be switched off and improve performance.

File: piq/iw_ssim.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+    sigma_xx = F.relu(sigma_xx)
+    sigma_yy = F.relu(sigma_yy)
+
+    # Contrast sensitivity (CS) with alpha = beta = gamma = 1.
+    cs_map = (2. * sigma_xy + c2) / (sigma_xx + sigma_yy + c2)
+
+    # Structural similarity (SSIM)
+    ss_map = (2. * mu_xy + c1) / (mu_xx + mu_yy + c1) * cs_map
+
+    return ss_map, cs_map
```
Comment (zakajd) [inline piq/iw_ssim.py:305]: What's the difference between this function and one defined in `ssim.py`? Can we reuse it?

File: piq/iw_ssim.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+                                              sigma_nsq * sigma_nsq)), dim=-1)
+
+    iw_map[iw_map < EPS] = 0
+
+    return iw_map
+
+
+def _image_enlarge(x: torch.Tensor) -> torch.Tensor:
+    r"""Bilinear upscaling of the image.
+
```
Comment (zakajd) [inline piq/iw_ssim.py:428]: This function does more than just image upscaling. Can you clarify what's happening in lines 436-440 in docstring?

File: piq/iw_ssim.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+    Args:
+        x: An input tensor. Shape :math:`(N, C, H, W)`.
+        shift: Offset list.
+
+    Returns:
+        The circular shiftet tensor.
+    """
+    tmp = torch.cat((x[..., -shift[0]:, :], x[..., :-shift[0], :]), dim=-2)
+    tmp = torch.cat((tmp[..., -shift[1]:], tmp[..., :-shift[1]]), dim=-1)
+    return tmp
```
Comment (zakajd) [inline piq/iw_ssim.py:458]: Maybe change variable name to something more meaningful? Like `x_shifted`?

**Review by zakajd — CHANGES_REQUESTED** [review]

File: tests/test_iw_ssim.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+
+@pytest.fixture(scope='module')
+def x() -> torch.Tensor:
+    return torch.rand(3, 3, 161, 161)
+
+
+@pytest.fixture(scope='module')
+def y() -> torch.Tensor:
+    return torch.rand(3, 3, 161, 161)
```
[the remark this answers is not in the document]
Comment (denproc) [reply tests/test_iw_ssim.py]: The common test tensors have size of `(4, 3, 96, 96)`. Unfortunately, IW-SSIM with default parameters cannot process images with resolution less than `161x161`. For default parameters we have four downscaling operations and convolution with kernel size `11`. Therefore, tensors with different shape are required

File: piq/iw_ssim.py [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+        Value of Information Content Weighted Structural Similarity (IW-SSIM) index.
+
+    References:
+        Wang, Zhou, and Qiang Li..
+        Information content weighting for perceptual image quality assessment.
+        IEEE Transactions on image processing 20.5 (2011): 1185-1198.
+        https://ece.uwaterloo.ca/~z70wang/publications/IWSSIM.pdf DOI:`10.1109/TIP.2010.2092435`
+
+    Note:
+        Lack of content in the image could lead to NaN values.
```
Comment (snk4tr) [inline piq/iw_ssim.py:61]: What exactly will be returned? The `None` value or `torch.tensor(['nan])`? I would strongly prefer to stick with the latter.

File: piq/iw_ssim.py [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+    """
+    assert kernel_size % 2 == 1, f'Kernel size must be odd, got [{kernel_size}]'
```
Comment (snk4tr) [inline piq/iw_ssim.py:63]: Shall we also validate here that `parent` values are valid (in some `(0, min_size] range)?

File: piq/ms_ssim.py [hunk 8]
```diff
@@ -69,6 +69,8 @@ def multi_scale_ssim(x: torch.Tensor, y: torch.Tensor, kernel_size: int = 11, ke
     else:
         # Normalize scale weights
         scale_weights = (scale_weights / scale_weights.sum()).to(x)
+    assert scale_weights.size(0) == scale_weights.numel(), \
+        f'Expected a vector of weights, got {scale_weights.dim()}D tensor'
```
Comment (snk4tr) [inline piq/ms_ssim.py:73]: :+1:

File: piq/iw_ssim.py [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+    if x.size(1) == 3:
+        x = rgb2yiq(x)[:, :1]
+        y = rgb2yiq(y)[:, :1]
+
+    if scale_weights is None:
+        scale_weights = torch.tensor([0.0448, 0.2856, 0.3001, 0.2363, 0.1333], dtype=x.dtype, device=x.device)
+    scale_weights = scale_weights / scale_weights.sum()
+    assert scale_weights.size(0) == scale_weights.numel(),\
+        f'Expected a vector of weights, got {scale_weights.dim()}D tensor'
+
```
Comment (snk4tr) [inline piq/iw_ssim.py:79]: Will  the code run with no exceptions until the end if this requirement is not satisfied? If no, I suggest raising an exception instead.

File: piq/iw_ssim.py [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+    levels = scale_weights.size(0)
+
+    min_size = (kernel_size - 1) * 2 ** (levels - 1) + 1
+    assert x.size(-1) >= min_size and x.size(-2) >= min_size,\
+        f'Invalid size of the input images, expected at least {min_size}x{min_size}.'
+
```
Comment (snk4tr) [inline piq/iw_ssim.py:85]: I think this one should remain an exception. Disabling it with `python -O` makes no sense, the code will fail later anyway. What we want to do with rising this exception is to make it fail early and save compute time.

File: piq/iw_ssim.py [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+    bound = math.ceil((kernel_size - 1) / 2)  # Ceil
+    bound1 = bound - math.floor((blk_size - 1) / 2)  # floor
```
Comment (snk4tr) [inline piq/iw_ssim.py:87]: Can this variable have a bit more descriptive name? :smile:
Or you can add a comment, otherwise it looks a bit confusing to me.

File: piq/iw_ssim.py [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+
+    C_u = torch.matmul(Y.transpose(-2, -1), Y) / nexp
+
+    recommended_torch_version = _parse_version('1.7.0')
+    torch_version = _parse_version(torch.__version__)
+    if len(torch_version) != 0 and torch_version >= recommended_torch_version:
+        eig_values, eig_vectors = torch.linalg.eigh(C_u)
+    else:
+        eig_values, eig_vectors = torch.symeig(C_u, eigenvectors=True)
+
```
Comment (snk4tr) [inline piq/iw_ssim.py:396]: Are you sure that it will work for all versions `<1.7.0`?

File: piq/iw_ssim.py [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+    """
+    t1 = F.interpolate(x, size=(int(4 * x.size(-2) - 3), int(4 * x.size(-1) - 3)), mode='bilinear', align_corners=False)
+    t2 = torch.zeros([x.size(0), 1, 4 * x.size(-2) - 1, 4 * x.size(-1) - 1]).to(x)
+    t2[:, :, 1: -1, 1:-1] = t1
+    t2[:, :, 0, :] = 2 * t2[:, :, 1, :] - t2[:, :, 2, :]
+    t2[:, :, -1, :] = 2 * t2[:, :, -2, :] - t2[:, :, -3, :]
+    t2[:, :, :, 0] = 2 * t2[:, :, :, 1] - t2[:, :, :, 2]
+    t2[:, :, :, -1] = 2 * t2[:, :, :, -2] - t2[:, :, :, -3]
+    out = t2[:, :, ::2, ::2]
+    return out
```
Comment (snk4tr) [inline piq/iw_ssim.py:444]: Wow! Happy times debugging it :smile:

File: tests/test_iw_ssim.py [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+    y_rgb = torch.tensor(imread('tests/assets/i01_01_5.bmp')).permute(2, 0, 1).unsqueeze(0)
+    return [(x_grey, y_grey), (x_rgb, y_rgb)]
+
+
+@pytest.fixture(scope='module')
+def scale_weights() -> torch.Tensor:
+    return torch.tensor([0.0448, 0.2856, 0.3001, 0.2363, 0.1333])
+
+
+# ================== Test function: `multi_scale_ssim` ==================
```
Comment (snk4tr) [inline tests/test_iw_ssim.py:44]: `multi_scale_ssim` -> `information_weighted_ssim`

Comment (snk4tr) [inline tests/test_iw_ssim.py:44]: `MultiScaleSSIMLoss` -> `InformationWeightedSSIMLoss`

**Review by snk4tr — CHANGES_REQUESTED** [review]
Great job @denproc. Only minor comments and questions

File: tests/test_iw_ssim.py [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+    assert torch.isclose(score_gray, matlab_gray, atol=1e-5),\
+        f'Expected {matlab_gray:.4f}, got {score_gray:.4f} for gray scale case.'
+
+    score_rgb = information_weighted_ssim(x_rgb.to(device), y_rgb.to(device), data_range=255)
+
+    assert torch.isclose(score_rgb, matlab_rgb, atol=1e-5),\
+        f'Expected {matlab_rgb:.8f}, got {score_rgb:.8f} for rgb case.'
+
+
+# ================== Test class: `MultiScaleSSIMLoss` ==================
```
Comment (snk4tr) [inline tests/test_iw_ssim.py:153]: `MultiScaleSSIMLoss` -> `InformationWeightedSSIMLoss`

File: piq/iw_ssim.py (hunk 7 above)
Comment (denproc) [reply piq/iw_ssim.py:63]: `parent` is a `bool` flag. Do you suggest to validate its type?

File: piq/iw_ssim.py (hunk 12 above)
Comment (denproc) [reply piq/iw_ssim.py:396]: It does work. Our GitHub pipeline checks it for `torch==[1.5.1, 1.8.1]`.

File: piq/iw_ssim.py (hunk 6 above)
Comment (denproc) [reply piq/iw_ssim.py]: NaN output was fixed. However, lack of content in target image could lead to singular matrix, which raises the exception during matrix inversion. I fixed the note according to current behaviour.

File: piq/iw_ssim.py (hunk 9 above)
Comment (denproc) [reply piq/iw_ssim.py:80]: `scale_weights` are used for the final operation to compute the score. The type of exception is related to a discussion in #305, where I suggest to continue discussion

File: piq/iw_ssim.py (hunk 10 above)
Comment (denproc) [reply piq/iw_ssim.py:86]: This one is related to the discussion in #305, where I suggest to continue.

**Comment by denproc** [comment]
Ready for another review :)

**Review by snk4tr — APPROVED** [review]
@denproc sorry for the long delay. Thank you for answering questions and incorporating changes. Merging!

**Review by zakajd — APPROVED** [review]
